core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,465 @@
+    // update occurs, the iterative logic will still be present for the first pass through for
+    // the new address list even though we are in a state of TRANSIENT_FAILURE.
+    if (currentState == TRANSIENT_FAILURE) {
+      if (newState == CONNECTING) {
+        // each subchannel is responsible for its own backoff
+        return;
+      } else if (newState == IDLE) {
+        requestConnection();
+      }
+    }
+
+    // update subchannel state unless it was in transient failurere
+    if (states.get(subchannel) != TRANSIENT_FAILURE || newState != CONNECTING) {
+      states.put(subchannel, newState);
+    }
+
+    switch (newState) {
+      case IDLE:
+        // Shutdown when ready: connect from beginning when prompted
+        addressIndex.reset();
+        updateBalancingState(IDLE, new RequestConnectionPicker(this));;
+        break;
+      case CONNECTING:
+        // It's safe to use RequestConnectionPicker here, so when coming from IDLE we could leave
+        // the current picker in-place. But ignoring the potential optimization is simpler.
+        updateBalancingState(CONNECTING, new Picker(PickResult.withNoResult()));
+        break;
+      case READY:
+        updateBalancingState(READY, new Picker(PickResult.withSubchannel(subchannel)));
+        shutdownRemaining(subchannel.getAddresses().getAddresses().get(0), subchannel);
+        addressIndex.reset();
+        break;
+      case TRANSIENT_FAILURE:
+        // If we are looking at current channel ...
+        if (subchannels.get(addressIndex.getCurrentAddress()).equals(subchannel)) {
+          // ... and it is the last address, report TRANSIENT_FAILURE.
+          // Only happens once per address update.
+          addressIndex.increment();
+          if (!addressIndex.isValid()) {
+            addressIndex.reset();
+            helper.refreshNameResolution();
+            updateBalancingState(TRANSIENT_FAILURE,
+                new Picker(PickResult.withError(stateInfo.getStatus())));
+            // If we are not at the last address, try a connection attempt to the next address.
+          } else {
+            requestConnection();
+            if (!addressIndex.isValid() && currentState != TRANSIENT_FAILURE) {
+              updateBalancingState(TRANSIENT_FAILURE,
+                  new Picker(PickResult.withError(stateInfo.getStatus())));
+            }
+          }
+          // If we are looking at another address that retried after backoff
+          // and entered TRANSIENT_FAILURE, we do not do any incrementing logic,
+          // as individual subchannels will perform their own backoffs.
+        }
+        break;
+      default:
+        throw new IllegalArgumentException("Unsupported state:" + newState);
+    }
+  }
+
+  private void updateBalancingState(ConnectivityState state, SubchannelPicker picker) {
+    if (state != currentState || currentState != CONNECTING) {
```

Comment:
Why `currentState!=CONNECTING`? We should do helper.updateBalancingState if state changes or picker changes.

Reply:
We talked about an optimization to not update states if we're already in CONNECTING and we're told to updateBalancingState to CONNECTING again. 

This is the same thing as `!(state == currentState && currentState == CONNECTING)`, but I just "simplified" it. I admit that it makes it more confusing, so should I change it to the format above?

Reply:
state != currentState || oldPicker != newPicker